lib/matplotlib/backend_bases.py

```diff
@@ -3195,3 +3207,560 @@ def zoom(self, *args):
+        self.canvas = canvas
+
+        self._key_press_handler_id = self.canvas.mpl_connect(
+            'key_press_event', self._key_press)
+
+        self._tools = {}
+        self._keys = {}
+        self._toggled = {}
+        self._callbacks = cbook.CallbackRegistry()
+
+        # to process keypress event
+        self.keypresslock = widgets.LockDraw()
+        self.messagelock = widgets.LockDraw()
+
+    def nav_connect(self, s, func):
+        """
+        Connect event with string *s* to *func*.
+
+        Parameters
+        -----------
+        s : String
+            Name of the event
+
+            The following events are recognized
+
+            - 'tool_message_event'
+            - 'tool_removed_event'
+            - 'tool_added_event'
+
+            For every tool added a new event is created
+
+            - 'tool_trigger_TOOLNAME`
+              Where TOOLNAME is the id of the tool.
+
+        func : function
+            Function to be called with signature
+            def func(event)
+        """
+        return self._callbacks.connect(s, func)
+
+    def nav_disconnect(self, cid):
+        """
+        Disconnect callback id cid
+
+        Example usage::
+
+            cid = navigation.nav_connect('tool_trigger_zoom', on_press)
+            #...later
+            navigation.nav_disconnect(cid)
+        """
+        return self._callbacks.disconnect(cid)
+
+    def message_event(self, message, sender=None):
+        """ Emit a `NavigationMessageEvent`"""
+        if sender is None:
+            sender = self
+
+        s = 'tool_message_event'
+        event = NavigationMessageEvent(s, sender, message)
+        self._callbacks.process(s, event)
+
+    @property
+    def active_toggle(self):
+        """
+        Toggled Tool
+
+        **dict** :  Currently toggled tools
+        """
+
+        return self._toggled
+
+    def get_tool_keymap(self, name):
+        """
+        Get the keymap associated with the specified tool
+
+        Parameters
+        ----------
+        name : string
+            Name of the Tool
+
+        Returns
+        ----------
+        list : list of keys associated with the Tool
+        """
+
+        keys = [k for k, i in six.iteritems(self._keys) if i == name]
+        return keys
+
+    def _remove_keys(self, name):
+        for k in self.get_tool_keymap(name):
+            del self._keys[k]
+
+    def set_tool_keymap(self, name, *keys):
+        """
+        Set the keymap to associate with the specified tool
+
+        Parameters
+        ----------
+        name : string
+            Name of the Tool
+        keys : keys to associate with the Tool
+        """
+
+        if name not in self._tools:
+            raise KeyError('%s not in Tools' % name)
+
+        self._remove_keys(name)
+
+        for key in keys:
+            for k in validate_stringlist(key):
+                if k in self._keys:
+                    warnings.warn('Key %s changed from %s to %s' %
```

Comment:
I understand your point, but imho, if you really care about a key, you can remap it back later, those who care would take the time to remap it back later, and look to the console to figure out what happened.

Alternatives: with an ``overwrite=False`` kwarg, I would have thought it better that if ``True`` it overwrites; if ``False`` it keeps the existing key, both with warning messages.  I think this sits happier with me because it gives an extra feature, rather than adding a show-stopper.

Alternative 2:  For a later PR, I think we can add a tool to change the keymaps from the UI.  After MEP27 I will come back to the subplots tool to make this dialog use the new ``WindowBase``, eliminating the need for separate backends.  If this remap tool needs a dialog then I will do that then.

Finally, about ``warnings.warn``, I have misgivings about the use of this:
1. It only shows a string once, i.e. here it will only report the first change of that this key got changed from toolA to toolB, annoying when trying to figure out what went happened during debug.
2. As I refactor ``webagg``, I realise that afaik, warning.warn would print to the terminal, i.e. the server, instead of the client, a.k.a user who may well exist as two different people.  The only problem here lies in that this opens up a whole new can of worms about messages (do we print to the GUI message area, do we print to terminal, do we print to both, do we need ``message-types`` to help determine that, what should they look like etcetera) so I feel this discussion should probably happen later, perhaps even in its own MEP.